### PR TITLE
ovndbmanager: Implement unit tests for missing functions

### LINKED ISSUE #2576: Unit testing: Add more coverage for ovn-dbchecker
As I was working on #2554 I noticed we have unit tests for the election timer check, but no other checks. It would be nice if someone can add testing for each of the other checks: ensureClusterRaftMembership, ensureLocalRaftServerID, as well as testing if the db would get reset. Can leverage the changes I'm making in to have the checks return an error.

### PR DESCRIPTION
**Pull request by andreaskaris**
Implement unit tests for ensureLocalRaftServerID,
ensureClusterRaftMembership, resetRaftDB and add minor
changes to these functions.

Fixes #2576

### CODE REVIEW

File: go-controller/pkg/ovndbmanager/ovndbmanager_test.go [hunk 1]
```diff
@@ -114,9 +494,110 @@ func TestElectionTimer(t *testing.T) {
...
+			for k, v := range tc.mockCalls {
+				mockCalls[k] = v
+			}
+
+			db.dbName = tc.dbName
+			db.dbAlias = tc.dbAlias
+
+			// resetRaftDb expects a file in the current directory with the db alias'
+			// name; create it
+			createDbFile(t, tc.dbAlias, tc.createDbFile)
```
Comment (dcbw) [inline go-controller/pkg/ovndbmanager/ovndbmanager_test.go:591]: @andreaskaris I would just do:
```
if tc.createDbFile {
    createDbFile(...)
}
```

File: go-controller/pkg/ovndbmanager/ovndbmanager_test.go [hunk 2]
```diff
@@ -132,3 +613,45 @@ func TestElectionTimer(t *testing.T) {
 func keyForArgs(args ...string) string {
 	return strings.Join(args, "-")
 }
+
+// create file with name, fail on error or if the file already exists
+func createDbFile(t *testing.T, name string, create bool) {
```
Comment (dcbw) [inline go-controller/pkg/ovndbmanager/ovndbmanager_test.go:618]: @andreaskaris best practice is to create a temp dir with t.TempDir() at the beginning of the test in which you can write files. That gets cleaned up for you automatically. So I would do that at the top of `TestResetRaftDB` and then pass that directory to `createDbFile()` and `filepath.Join()` the dbname to that directory.

File: go-controller/pkg/ovndbmanager/ovndbmanager_test.go (hunk 1 above)
Comment (andreaskaris) [reply go-controller/pkg/ovndbmanager/ovndbmanager_test.go]: Done

File: go-controller/pkg/ovndbmanager/ovndbmanager_test.go (hunk 2 above)
Comment (andreaskaris) [reply go-controller/pkg/ovndbmanager/ovndbmanager_test.go]: I added this to the top of the Test function:
~~~
func TestResetRaftDB(t *testing.T) {
	// resetRaftDB expects a file named ovnnb or ovnsb in the local directory.
	// Change to a temporary directory for this test and switch back once the test returns.
	tmpDir := t.TempDir()
	currentDir, err := os.Getwd()
	if err != nil {
		t.Fatalf("Cannot get current directory: %s", err.Error())
	}
	err = os.Chdir(tmpDir)
	if err != nil {
		t.Fatalf("Cannot change to temporaryr directory: %s", err.Error())
	}
	t.Logf("Switched to temporary directory: %s", tmpDir)
	defer func() {
		err := os.Chdir(currentDir)
		if err != nil {
			t.Fatalf("Cannot change back to current directory: %s", err.Error())
		}
		t.Logf("Switched back to current directory: %s", currentDir)
	}()
~~~

The beginning of resetRaftDb looks like this:
~~~
// resetRaftDB backs up the db by renaming it and then stops the nb/sb ovsdb process.
// Returns an error if anything goes wrong.
func resetRaftDB(db *dbProperties) error {
        dbFile := filepath.Base(db.dbAlias)
        backupFile := strings.TrimSuffix(dbFile, filepath.Ext(dbFile)) +
                time.Now().UTC().Format("2006-01-02_150405") + "db_bak"
        backupDB := filepath.Join(filepath.Dir(db.dbAlias), backupFile)
        err := os.Rename(db.dbAlias, backupDB)
        if err != nil {
                return fmt.Errorf("failed to back up the db to backupFile: %s", backupFile)
        }
~~~

So I think that I either have to os.Chdir to the temp directory (as in the current code suggestion) or I have to modify the production code further and add a variable which holds the current directory?

**Comment by andreaskaris** [comment]
@dcbw Great thanks for the review. Can we merge this in? The fail is due to the issue that Surya is working on :-)

File: go-controller/pkg/ovndbmanager/ovndbmanager_test.go [hunk 3]
```diff
@@ -132,3 +618,31 @@ func TestElectionTimer(t *testing.T) {
...
+// failOnErrorMismatch fails either if an error is seen but not expected
+// or if an error is expected but when the substring does not match the error
+func failOnErrorMismatch(t *testing.T, err error, errorString string) {
+	if err != nil {
+		if errorString == "" || !strings.Contains(err.Error(), errorString) {
+			t.Errorf("Unexpected outcome. Got error '%v' but it does not contain expected string '%s'", err, errorString)
+		}
+	} else {
+		if errorString != "" {
+			t.Errorf("No error expected. However, received '%v'", err)
```
Comment (ricky-rav) [inline go-controller/pkg/ovndbmanager/ovndbmanager_test.go:645]: this looks to me more like we expected an error (`errorString != ""`) but got none (`err == nil`), right?

File: go-controller/pkg/ovndbmanager/ovndbmanager_test.go [hunk 4]
```diff
@@ -132,3 +618,31 @@ func TestElectionTimer(t *testing.T) {
...
 	return strings.Join(args, "-")
 }
+
+// create file with name, fail on error or if the file already exists
+func createDbFile(t *testing.T, name string) {
+	_, err := os.Stat(name)
+	if os.IsNotExist(err) {
+		f, err := os.OpenFile(name, os.O_RDONLY|os.O_CREATE, 0644)
+		if err != nil {
+			t.Errorf("Unexpected errror: %v", err)
```
Comment (ricky-rav) [inline go-controller/pkg/ovndbmanager/ovndbmanager_test.go:628]: too many r's in "error" :)

File: go-controller/pkg/ovndbmanager/ovndbmanager_test.go (hunk 3 above)
Comment (andreaskaris) [reply go-controller/pkg/ovndbmanager/ovndbmanager_test.go:645]: Yes. I wanted to get rid of some duplicate logic inside the test methods.

E.g. the test case looks like this:
~~~
		{
			desc: "Test error: failed to change election timer when leader timer must change but desired is more than double",
			mockCalls: map[string]*mockRes{
				keyForArgs("cluster/change-election-timer", "OVN_Northbound", "3000"): {
					res:    "",
					stderr: "failure",
					err:    fmt.Errorf("failure"),
				},
			},
			timeout:      5000,
			role:         "leader",
			currentTimer: "1500",
			errorString:  "failed to change election timer for",  // <---------- expected err string
		},
~~~

And the call to failOnErrorMismatch looks like this:
~~~
		// fail either if an error is seen but not expected
			// or if an error is expected but when the substring does not match the error
			failOnErrorMismatch(t, err, tc.errorString)
~~~

And method failOnErrorMismatch:
~~~
// failOnErrorMismatch fails either if an error is seen but not expected
// or if an error is expected but when the substring does not match the error
func failOnErrorMismatch(t *testing.T, err error, errorString string) {
	if err != nil {
		if errorString == "" || !strings.Contains(err.Error(), errorString) {
			t.Errorf("Unexpected outcome. Got error '%v' but it does not contain expected string '%s'", err, errorString)
		}
	} else {
		if errorString != "" {
			t.Errorf("No error expected. However, received '%v'", err)	
		}
	}
}
~~~

So we fail on a mismatch between err (returned from the function under test) and errorString (provided in the test cases).

So in other words, if err == nil and errorString != "", then this is invalid.
if err != nil and if errorString == "" or if the errorString does not match the errors Error message, then this is invalid, too.
 
I might rewrite the `} else { if errorString != "" {` but I think the current version is easier on the eyes.

WDYT?

File: go-controller/pkg/ovndbmanager/ovndbmanager_test.go (hunk 4 above)
Comment (andreaskaris) [reply go-controller/pkg/ovndbmanager/ovndbmanager_test.go]: Oh thanks! Resolved.

File: go-controller/pkg/ovndbmanager/ovndbmanager_test.go (hunk 3 above)
Comment (ricky-rav) [reply go-controller/pkg/ovndbmanager/ovndbmanager_test.go:645]: It's very well done, only the error message that I was quoting - if I followed everything correctly - should instead be reversed: "Expected error xyz, but got none", right?

Comment (andreaskaris) [reply go-controller/pkg/ovndbmanager/ovndbmanager_test.go]: Indeed. I changed this to:
~~~
// failOnErrorMismatch fails either if an error is seen but not expected
// or if an error is expected but when the substring does not match the error
func failOnErrorMismatch(t *testing.T, receivedErr error, expectedErrorString string) {
	if receivedErr != nil {
		if expectedErrorString == "" {
			t.Errorf("No error expected. However, received '%v' from method under test.", receivedErr)
		} else if !strings.Contains(receivedErr.Error(), expectedErrorString) {
			t.Errorf("Expected error string to contain '%s'. However, method under test threw error '%v'.", expectedErrorString, receivedErr)
		}
	} else if expectedErrorString != "" {
		t.Errorf("Error with error string '%s' expected. However, method under test completed without error.", expectedErrorString)
	}
}
~~~

I also noticed that existing error messages:
~~~
			return fmt.Errorf("failed to change election timer for %s %v %v", db.dbName, err, stderr)
(...)
                        return fmt.Errorf("failed to change election timer for %s %v %v", db.dbName, err, stderr)
~~~

would return "failed to change election timeout for ... failure failure". I changed that to include commas, e.g.:
~~~
return fmt.Errorf("failed to change election timer for %s, %v, %v", db.dbName, err, stderr)
~~~

So that it's a bit clearer that one is err, and the other one stderr.

File: go-controller/pkg/ovndbmanager/ovndbmanager.go [hunk 5]
```diff
@@ -273,12 +262,12 @@ func ensureClusterRaftMembership(db string, kclient kube.Interface) error {
...
 			// unknown member and we have enough members its safe to kick the unknown address
-			klog.Infof("Unknown Raft member found in %s: %s, %s... kicking", db, member[1], member[3])
-			_, stderr, err = appCtl(5, "cluster/kick", dbName, member[1])
+			klog.Infof("Unknown Raft member found in %s: %s, %s... kicking", db.dbName, member[1], member[3])
+			_, stderr, err = db.appCtl(5, "cluster/kick", db.dbName, member[1])
 			if err != nil {
 				// warn only: we might fail to kick since other nodes will also be trying to kick the member
 				klog.Warningf("Error while kicking old Raft member: %s, for address: %s in db: %s,"+
-					"stderr: %v, err: %v", member[1], member[3], db, stderr, err)
+					"stderr: %v, err: %v", member[1], member[2], db.dbAlias, stderr, err)
```
Comment (trozet) [inline go-controller/pkg/ovndbmanager/ovndbmanager.go:270]: @andreaskaris is it intentional here to change member[3] to member[2] ?

Comment (andreaskaris) [reply go-controller/pkg/ovndbmanager/ovndbmanager.go]: Nope. Typo. me~~ ; I fixed it and made the tests run into this path of the code now, too:
 https://github.com/ovn-org/ovn-kubernetes/pull/2582/commits/8181b0ae2f5fae8b15f8524ed91e8fdf8f1c3a77#diff-6eecb13689493869d78a0b346a91937e693d3d4c97976b1d2aa431c964590656R308
That doesn't change anything to the outcome, but the message will be printed when running the tests.

Comment (andreaskaris) [reply go-controller/pkg/ovndbmanager/ovndbmanager.go]: [root@ovnkubernetes go-controller]# go test -count 1 -mod=vendor -test.v -race  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovndbmanager
I0105 11:42:33.961490  104349 ovs.go:93] Maximum command line arguments set to: 191102
=== RUN   TestEnsureLocalRaftServerID
=== RUN   TestEnsureLocalRaftServerID/0:Test_error:_unable_to_get_db_server_ID
=== RUN   TestEnsureLocalRaftServerID/1:Test_error:_Invalid_sid
=== RUN   TestEnsureLocalRaftServerID/2:Test_error:_Unable_to_get_cluster_status
=== RUN   TestEnsureLocalRaftServerID/3:Test_error:_unable_to_parse_address_for_db
=== RUN   TestEnsureLocalRaftServerID/4:Test_error:_while_kicking_old_Raft_member
I0105 11:42:33.970319  104349 ovndbmanager.go:178] Previous stale member found in OVN_Northbound: 3936... kicking
=== RUN   TestEnsureLocalRaftServerID/5:Stale_member,_kick
I0105 11:42:33.971577  104349 ovndbmanager.go:178] Previous stale member found in OVN_Northbound: 3936... kicking
=== RUN   TestEnsureLocalRaftServerID/6:Consistent_database,_no_action_needed
--- PASS: TestEnsureLocalRaftServerID (0.00s)
    --- PASS: TestEnsureLocalRaftServerID/0:Test_error:_unable_to_get_db_server_ID (0.00s)
    --- PASS: TestEnsureLocalRaftServerID/1:Test_error:_Invalid_sid (0.00s)
    --- PASS: TestEnsureLocalRaftServerID/2:Test_error:_Unable_to_get_cluster_status (0.00s)
    --- PASS: TestEnsureLocalRaftServerID/3:Test_error:_unable_to_parse_address_for_db (0.00s)
    --- PASS: TestEnsureLocalRaftServerID/4:Test_error:_while_kicking_old_Raft_member (0.00s)
    --- PASS: TestEnsureLocalRaftServerID/5:Stale_member,_kick (0.00s)
    --- PASS: TestEnsureLocalRaftServerID/6:Consistent_database,_no_action_needed (0.00s)
=== RUN   TestEnsureClusterRaftMembership
=== RUN   TestEnsureClusterRaftMembership/0:Test_error:_Invalid_database_name
=== RUN   TestEnsureClusterRaftMembership/1:Test_error:_Unable_to_get_cluster_status
=== RUN   TestEnsureClusterRaftMembership/2:Consistent_database,_no_action_needed
**=== RUN   TestEnsureClusterRaftMembership/3:Unknown_Raft_member,_kick**
**I0105 11:42:33.977329  104349 ovndbmanager.go:265] Unknown Raft member found in OVN_Northbound: c10c, 10.1.1.219:9643... kicking**
**I0105 11:42:33.977382  104349 ovndbmanager.go:265] Unknown Raft member found in OVN_Northbound: fc43, 10.1.1.220:9643... kicking**
**W0105 11:42:33.977408  104349 ovndbmanager.go:269] Error while kicking old Raft member: fc43, for address: 10.1.1.220:9643 in db: ovnnb,stderr: unknown server**
**ovn-appctl: /var/run/ovn/ovnnb_db.ctl: server returned an error, err: Error unkown server**
**--- PASS: TestEnsureClusterRaftMembership (0.00s)**
    --- PASS: TestEnsureClusterRaftMembership/0:Test_error:_Invalid_database_name (0.00s)
    --- PASS: TestEnsureClusterRaftMembership/1:Test_error:_Unable_to_get_cluster_status (0.00s)
    --- PASS: TestEnsureClusterRaftMembership/2:Consistent_database,_no_action_needed (0.00s)
**    --- PASS: TestEnsureClusterRaftMembership/3:Unknown_Raft_member,_kick (0.00s)**
=== RUN   TestEnsureElectionTimeout
=== RUN   TestEnsureElectionTimeout/0:Test_error:_Unable_to_get_cluster_status
=== RUN   TestEnsureElectionTimeout/1:Test_error:_Failed_to_get_current_election_timer
=== RUN   TestEnsureElectionTimeout/2:Follower,_not_trying_to_change
=== RUN   TestEnsureElectionTimeout/3:leader,_timer_doesn't_change
=== RUN   TestEnsureElectionTimeout/4:Test_error:_failed_to_change_election_timer_when_leader_timer_must_change
=== RUN   TestEnsureElectionTimeout/5:Test_error:_failed_to_change_election_timer_when_leader_timer_must_change_but_desired_is_more_than_double
=== RUN   TestEnsureElectionTimeout/6:leader,_timer_must_change
=== RUN   TestEnsureElectionTimeout/7:leader,_timer_must_change_but_desired_is_more_than_double
--- PASS: TestEnsureElectionTimeout (0.00s)
    --- PASS: TestEnsureElectionTimeout/0:Test_error:_Unable_to_get_cluster_status (0.00s)
    --- PASS: TestEnsureElectionTimeout/1:Test_error:_Failed_to_get_current_election_timer (0.00s)
    --- PASS: TestEnsureElectionTimeout/2:Follower,_not_trying_to_change (0.00s)
    --- PASS: TestEnsureElectionTimeout/3:leader,_timer_doesn't_change (0.00s)
    --- PASS: TestEnsureElectionTimeout/4:Test_error:_failed_to_change_election_timer_when_leader_timer_must_change (0.00s)
    --- PASS: TestEnsureElectionTimeout/5:Test_error:_failed_to_change_election_timer_when_leader_timer_must_change_but_desired_is_more_than_double (0.00s)
    --- PASS: TestEnsureElectionTimeout/6:leader,_timer_must_change (0.00s)
    --- PASS: TestEnsureElectionTimeout/7:leader,_timer_must_change_but_desired_is_more_than_double (0.00s)
=== RUN   TestResetRaftDB
=== RUN   TestResetRaftDB/0:Failed_to_back_up_the_db_to_backupFile
=== RUN   TestResetRaftDB/1:Failed_to_restart_the_database
I0105 11:42:33.981277  104349 ovndbmanager.go:331] Backed up the db to backupFile: ovnnb2022-01-05_114233db_bak
=== RUN   TestResetRaftDB/2:Successful_database_backup
I0105 11:42:33.981729  104349 ovndbmanager.go:331] Backed up the db to backupFile: ovnnb2022-01-05_114233db_bak
I0105 11:42:33.981759  104349 ovndbmanager.go:337] Stopped OVN_Northbound db after backing up the db: ovnnb2022-01-05_114233db_bak
--- PASS: TestResetRaftDB (0.00s)
    --- PASS: TestResetRaftDB/0:Failed_to_back_up_the_db_to_backupFile (0.00s)
    --- PASS: TestResetRaftDB/1:Failed_to_restart_the_database (0.00s)
    --- PASS: TestResetRaftDB/2:Successful_database_backup (0.00s)
PASS
ok  	github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovndbmanager	0.401s

**Review by trozet — APPROVED** [review]
